opds2_import.py

```diff
@@ -0,0 +1,830 @@
+                encode(link), encode(media_types_and_drm_scheme)
+            )
+        )
+
+        return media_types_and_drm_scheme
+
+    def _extract_medium_from_links(self, links):
+        """Extract the publication's medium from its links.
+
+        :param links: List of links
+        :type links: ast_core.LinkList
+
+        :return: Publication's medium
+        :rtype: Optional[str]
+        """
+        derived = None
+
+        for link in links:
+            if not link.rels or not link.type or not self._is_acquisition_link(link):
+                continue
+
+            link_media_type, _ = first_or_default(
+                self._extract_media_types_and_drm_scheme_from_link(link),
+                default=(None, None),
+            )
+            derived = Edition.medium_from_media_type(link_media_type)
+
+            if derived:
+                break
+
+        return derived
+
+    @staticmethod
+    def _extract_medium(publication, default_medium=Edition.BOOK_MEDIUM):
+        """Extract the publication's medium from its metadata.
+
+        :param publication: Publication object
+        :type publication: opds2_core.OPDS2Publication
+
+        :return: Publication's medium
+        :rtype: str
+        """
+        medium = default_medium
+
+        if publication.metadata.type:
+            medium = Edition.additional_type_to_medium.get(
+                publication.metadata.type, default_medium
+            )
+
+        return medium
+
+    def _extract_publication_metadata(self, feed, publication, data_source_name):
+        """Extract a Metadata object from webpub-manifest-parser's publication.
+
+        :param publication: Feed object
+        :type publication: opds2_ast.OPDS2Feed
+
+        :param publication: Publication object
+        :type publication: opds2_ast.OPDS2Publication
+
+        :param data_source_name: Data source's name
+        :type data_source_name: str
+
+        :return: Publication's metadata
+        :rtype: Metadata
+        """
+        self._logger.debug(
+            u"Started extracting metadata from publication {0}".format(
+                encode(publication)
+            )
+        )
+
+        title = publication.metadata.title
+
+        if title == OPDSFeed.NO_TITLE:
+            title = None
+
+        subtitle = publication.metadata.subtitle
+
+        languages = first_or_default(publication.metadata.languages)
+        derived_medium = self._extract_medium_from_links(publication.links)
+        medium = self._extract_medium(publication, derived_medium)
+
+        publisher = first_or_default(publication.metadata.publishers)
+        if publisher:
+            publisher = publisher.name
+
+        published = publication.metadata.published
+        subjects = self._extract_subjects(publication.metadata.subjects)
+        contributors = (
+            self._extract_contributors(publication.metadata.authors)
```

Comment:
Whenever you call _extract_contributors you should pass in a role constant corresponding to the part of the metadata you're extracting.

Reply:
Added `default_role` parameter to `_extract_contributors`